### ISSUE TITLE
Issues with rate limit + batching?

### ISSUE THREAD

**Issue by matt1484** [benthosdev/benthos#1169]
so I have a config of this general format: 
```yaml
rate_limit_resources:
  - label: no_spam
    local:
      count: 1
      interval: 100ms
input:
  processors:
    - rate_limit:
        resource: no_spam
  kafka:
    batching:
      count: 200
      period: 1s
      processors:
        - log:
            level: INFO
            message: 'pulling ${! batch_size() } messages from kafka'
output:
  broker:
    pattern: fan_out_sequential
    outputs:
      - elasticsearch: {}
        processors:
          - log:
              level: INFO
              message: 'pushing ${! batch_size() } messages to ES'
      - kafka: {}
        processors: 
          - log:
              level: INFO
              message: 'pushing ${! batch_size() } messages to kafka'
```
where I am attempting to get messages from kafka in batches of ~200 and push them through my outputs at a rate of 10 batches a second, but for some reason it eventually stops pulling in batches of 200 and appears to inefficiently always do batches of size 1 at a more frequent interval than expected. What is extra peculiar is that while monitoring kafka itself, the number of messages is going down at a faster rate than the logs (below) would indicate. Is this potentially a bug or do I have this misconfigured?

logs:
```
{"@timestamp":"2022-03-10T22:45:22Z","@service":"benthos","component":"benthos.input.0","level":"INFO","message":"pulled 200 messages from kafka"}
# repeated ~30 times
{"@timestamp":"2022-03-10T22:45:23Z","@service":"benthos","component":"benthos.input.0","level":"INFO","message":"pulled 4 messages from kafka"}
{"@timestamp":"2022-03-10T22:45:23Z","@service":"benthos","component":"benthos.input.0","level":"INFO","message":"pulled 9 messages from kafka"}
{"@timestamp":"2022-03-10T22:45:23Z","@service":"benthos","component":"benthos.input.0","level":"INFO","message":"pulled 3 messages from kafka"}
{"@timestamp":"2022-03-10T22:45:23Z","@service":"benthos","component":"benthos.input.0","level":"INFO","message":"pulled 4 messages from kafka"}
{"@timestamp":"2022-03-10T22:45:23Z","@service":"benthos","component":"benthos.input.0","level":"INFO","message":"pulled 3 messages from kafka"}
# random numbers pulled but eventually its always just one
{"@timestamp":"2022-03-10T22:45:32Z","@service":"benthos","component":"benthos.output.broker.outputs.0.fallback.0.processor.0","level":"INFO","message":"pushing 200 messages to ES"}
{"@timestamp":"2022-03-10T22:45:33Z","@service":"benthos","component":"benthos.output.broker.outputs.1.fallback.0.processor.1","level":"INFO","message":"pushing 200 messages to kafka"}
{"@timestamp":"2022-03-10T22:45:33Z","@service":"benthos","component":"benthos.input.0","level":"INFO","message":"pulled 1 messages from kafka"}
# process messages until eventually it becomes a spam like this (these are NOT copies of the same lines, thats the actual log)...
{"@timestamp":"2022-03-10T22:50:31Z","@service":"benthos","component":"benthos.output.broker.outputs.0.fallback.0.processor.0","level":"INFO","message":"pushing 1 messages to ES"}
{"@timestamp":"2022-03-10T22:50:31Z","@service":"benthos","component":"benthos.output.broker.outputs.1.fallback.0.processor.1","level":"INFO","message":"pushing 1 messages to kafka"}
{"@timestamp":"2022-03-10T22:50:31Z","@service":"benthos","component":"benthos.input.0","level":"INFO","message":"pulled 1 messages from kafka"}
{"@timestamp":"2022-03-10T22:50:31Z","@service":"benthos","component":"benthos.output.broker.outputs.0.fallback.0.processor.0","level":"INFO","message":"pushing 1 messages to ES"}
{"@timestamp":"2022-03-10T22:50:31Z","@service":"benthos","component":"benthos.output.broker.outputs.1.fallback.0.processor.1","level":"INFO","message":"pushing 1 messages to kafka"}
{"@timestamp":"2022-03-10T22:50:31Z","@service":"benthos","component":"benthos.input.0","level":"INFO","message":"pulled 1 messages from kafka"}
{"@timestamp":"2022-03-10T22:50:31Z","@service":"benthos","component":"benthos.output.broker.outputs.0.fallback.0.processor.0","level":"INFO","message":"pushing 1 messages to ES"}
{"@timestamp":"2022-03-10T22:50:31Z","@service":"benthos","component":"benthos.output.broker.outputs.1.fallback.0.processor.1","level":"INFO","message":"pushing 1 messages to kafka"}
{"@timestamp":"2022-03-10T22:50:31Z","@service":"benthos","component":"benthos.input.0","level":"INFO","message":"pulled 1 messages from kafka"}
{"@timestamp":"2022-03-10T22:50:31Z","@service":"benthos","component":"benthos.output.broker.outputs.0.fallback.0.processor.0","level":"INFO","message":"pushing 1 messages to ES"}
{"@timestamp":"2022-03-10T22:50:31Z","@service":"benthos","component":"benthos.output.broker.outputs.1.fallback.0.processor.1","level":"INFO","message":"pushing 1 messages to kafka"}
{"@timestamp":"2022-03-10T22:50:31Z","@service":"benthos","component":"benthos.input.0","level":"INFO","message":"pulled 1 messages from kafka"}
```

**Comment by Jeffail** [comment]
Hey @matt1484, you're hitting an interesting interaction here between the kafka input consumers and the batching mechanisms, I'm pretty sure you can configure around this but I'll explain what's happening and maybe there's a place in the docs for this stuff to live permanently.

Firstly, the batching mechanism period trigger is timed from the last batch dispatched rather than the first message entering the next pending batch, so if the mechanism is left dry for >period time the first message entered will trigger it immediately. I don't really think there's a definitively right or wrong way to go but it'd be nice to configure batching policies to switch between these modes.

Secondly, kafka consumers (one for each topic/partition pair) operate in absolute transactional lock-step, which means they populate the batcher with a series of ordered messages (offset 1, followed by offset 2, and so on) and once the batcher triggers the messages are flushed downstream (into your rate limit processor). After that flush the consumer waits until the batch has been acknowledged downstream before beginning to populate the batcher again.

This is a bit of an odd behaviour but it also guarantees message ordering during processing and outputting data, as data that's blocked on back pressure or an unresponsive output can't be overtaken by subsequent messages taking a different swimlane.

So with that in mind what I suspect is happening here is:
- Kafka topic/partition consumer begins populating the batcher, hits the count limit and flushes the batch
- Batch enters rate limit processor, which blocks it beyond the batching period
- Kafka consumer blocks until batch is flushed downstream
- Kafka consumer adds first message to batcher, which triggers it due to the period specified
- Rinse and repeat

There is a way to disable the kafka consumer lock stepping which is the `checkpoint_limit` field: 
https://www.benthos.dev/docs/components/inputs/kafka#checkpoint_limit. That determines how many messages we're willing to flush from a topic/partition without having commited the offsets. I originally expected ordering to be a hard requirement for many kafka users but the opposite is the reality. This is why in V4 I'm planning to change the default of `checkpoint_limit` so we avoid these lock-step issues.

I'm not sure whether it was your intention to rate limit based on individual messages of the batch or for batches aggregate, but if it's the latter you can also place the rate limit within a branch that eliminates all but one message and that'll mean it only counts for one action, so your config ends up something like this:

```yaml
rate_limit_resources:
  - label: no_spam
    local:
      count: 1
      interval: 100ms
input:
  processors:
    - branch:
        request_map: 'root = if batch_index() != 0 { deleted() }'
        processors:
          - rate_limit:
              resource: no_spam
  kafka:
    checkpoint_limit: 1000
    batching:
      count: 200
      period: 1s
      processors:
        - log:
            level: INFO
            message: 'pulling ${! batch_size() } messages from kafka'
output:
  broker:
    pattern: fan_out_sequential
    outputs:
      - elasticsearch: {}
        processors:
          - log:
              level: INFO
              message: 'pushing ${! batch_size() } messages to ES'
      - kafka: {}
        processors: 
          - log:
              level: INFO
              message: 'pushing ${! batch_size() } messages to kafka'
```

**Comment by matt1484** [comment]
so I tried your above suggestions with no such luck, but then I tried moving the rate limit processor to `kafka.batching.processors` and it works like 50% of the time which is super peculiar. It still ends up with a lot of 1s but occasionally there will be some 200s sprinkled throughout. Im not too sure what causes this though. Is it possible to get more consistency by adjusting both `fetch_buffer_cap` and `checkpoint_limit`. I only ask this because it seems that for my use case it may be easier to follow if I can more or less mimic a synchronous loop where I dont get new things until after I have pushed the current items, but Im not sure the best way to go about that
